ipamutils/utils.go

```diff
@@ -13,18 +16,33 @@ var (
 	// PredefinedGranularNetworks contains a list of 64K IPv4 private networks with host size 8
 	// (10.x.x.x/24) which do not overlap with the networks in `PredefinedBroadNetworks`
 	PredefinedGranularNetworks []*net.IPNet
-
-	initNetworksOnce sync.Once
+	initNetworksOnce           sync.Once
 )
 
-// InitNetworks initializes the pre-defined networks used by the built-in IP allocator
-func InitNetworks() {
+// PredefinedPools represent a set of address pools with prefix length Size.
+// Each pool in the set is derived from the Base pool. Base is to be passed
+// in CIDR format.Currently we support only local scope networks
+// Example: a Base "10.10.0.0/16 with Size 24 will define the set of 256
+// 10.10.[0-255].0/24 address pools
+type PredefinedPools struct {
+	Base string `json:"base"`
+	Size int    `json:"size"`
+}
+
+// InitNetworks initializes the local predefined address pools
+// with the default values.
+func InitNetworks(defaultAddressPool []*PredefinedPools) {
 	initNetworksOnce.Do(func() {
+		if defaultAddressPool != nil {
+			if err := InitAddressPools(defaultAddressPool); err != nil {
+				logrus.WithError(err).Error("InitAddressPools failed to initialize the default address pool")
+			}
+			return
+		}
 		PredefinedBroadNetworks = initBroadPredefinedNetworks()
 		PredefinedGranularNetworks = initGranularPredefinedNetworks()
```

Comment:
I think this most recent refactorization changes the behavior of the code from previous versions.  Previously, if the default address pool was not nil, the PredefinedGranularNetworks would still be initialized.  However, this code now does not initialize said pool if the default address pools are non-nil, even if InitAddressPool() does not return an error.  Is this intended?  Is the defaultAddressPool now supposed to supersede both Broad and Granular network pools?